### ISSUE TITLE
[!] Failed to download 'profile': can't modify frozen String

### ISSUE THREAD

**Issue by Buju77** [CocoaPods/CocoaPods#10920]
* [x] I've read and understood the [*CONTRIBUTING* guidelines and have done my best effort to follow](https://github.com/CocoaPods/CocoaPods/blob/master/CONTRIBUTING.md).

# Report

## What did you do?

Integrated one of our own hosted pod using the branch name:

`pod 'profile', git: 'ssh://git@blub.com:7999/profile.git', branch: 'epic/activities'`

and run the command normally:

`pod update --clean-install --verbose`

## What did you expect to happen?

Install all pod dependencies correctly.

## What happened instead?

Pod update failed during 'Fetching external sources' phase with `can't modify frozen String`. ❌ 

But `pod update` does work if I use the `commit: '...'` option instead of `branch: '...'`. ✅ 

Using the branch option also works fine using Cocoapods 1.9.1. This error only happened after upgrading to CP 1.10.2 now (we only upgraded recently).

## CocoaPods Environment

Cocoapods 1.10.2
Xcode 12.5.1
macOS 11.5
Ruby 2.6.3p62 (2019-04-16 revision 67580)

## Error

```
Fetching external sources
-> Pre-downloading: 'profile' from `ssh://git@blub.com:7999/profile.git`, branch `epic/activities`
  $ /usr/bin/git ls-remote ssh://git@blub.com:7999/profile.git epic/activities
  48f4b0a9e4241b2a93087dabca405f19ca239ea3	refs/heads/epic/activities
[!] Failed to download 'profile': can't modify frozen String

/Users/cloud/.rvm/gems/ruby-2.6.3/gems/cocoapods-1.10.2/lib/cocoapods/external_sources/abstract_external_source.rb:120:in `rescue in block in pre_download'
/Users/cloud/.rvm/gems/ruby-2.6.3/gems/cocoapods-1.10.2/lib/cocoapods/external_sources/abstract_external_source.rb:115:in `block in pre_download'
/Users/cloud/.rvm/gems/ruby-2.6.3/gems/cocoapods-1.10.2/lib/cocoapods/user_interface.rb:86:in `titled_section'
/Users/cloud/.rvm/gems/ruby-2.6.3/gems/cocoapods-1.10.2/lib/cocoapods/external_sources/abstract_external_source.rb:113:in `pre_download'
/Users/cloud/.rvm/gems/ruby-2.6.3/gems/cocoapods-1.10.2/lib/cocoapods/external_sources/downloader_source.rb:13:in `fetch'
/Users/cloud/.rvm/gems/ruby-2.6.3/gems/cocoapods-1.10.2/lib/cocoapods/installer/analyzer.rb:989:in `fetch_external_source'
/Users/cloud/.rvm/gems/ruby-2.6.3/gems/cocoapods-1.10.2/lib/cocoapods/installer/analyzer.rb:968:in `block (2 levels) in fetch_external_sources'
/Users/cloud/.rvm/gems/ruby-2.6.3/gems/cocoapods-1.10.2/lib/cocoapods/installer/analyzer.rb:967:in `each'
/Users/cloud/.rvm/gems/ruby-2.6.3/gems/cocoapods-1.10.2/lib/cocoapods/installer/analyzer.rb:967:in `block in fetch_external_sources'
/Users/cloud/.rvm/gems/ruby-2.6.3/gems/cocoapods-1.10.2/lib/cocoapods/user_interface.rb:64:in `section'
/Users/cloud/.rvm/gems/ruby-2.6.3/gems/cocoapods-1.10.2/lib/cocoapods/installer/analyzer.rb:966:in `fetch_external_sources'
/Users/cloud/.rvm/gems/ruby-2.6.3/gems/cocoapods-1.10.2/lib/cocoapods/installer/analyzer.rb:117:in `analyze'
/Users/cloud/.rvm/gems/ruby-2.6.3/gems/cocoapods-1.10.2/lib/cocoapods/installer.rb:414:in `analyze'
/Users/cloud/.rvm/gems/ruby-2.6.3/gems/cocoapods-1.10.2/lib/cocoapods/installer.rb:239:in `block in resolve_dependencies'
/Users/cloud/.rvm/gems/ruby-2.6.3/gems/cocoapods-1.10.2/lib/cocoapods/user_interface.rb:64:in `section'
/Users/cloud/.rvm/gems/ruby-2.6.3/gems/cocoapods-1.10.2/lib/cocoapods/installer.rb:238:in `resolve_dependencies'
/Users/cloud/.rvm/gems/ruby-2.6.3/gems/cocoapods-1.10.2/lib/cocoapods/installer.rb:160:in `install!'
/Users/cloud/.rvm/gems/ruby-2.6.3/gems/cocoapods-1.10.2/lib/cocoapods/command/update.rb:63:in `run'
/Users/cloud/.rvm/gems/ruby-2.6.3/gems/claide-1.0.3/lib/claide/command.rb:334:in `run'
/Users/cloud/.rvm/gems/ruby-2.6.3/gems/cocoapods-1.10.2/lib/cocoapods/command.rb:52:in `run'
/Users/cloud/.rvm/gems/ruby-2.6.3/gems/cocoapods-1.10.2/bin/pod:55:in `<top (required)>'
/Users/cloud/.rvm/gems/ruby-2.6.3/bin/pod:23:in `load'
/Users/cloud/.rvm/gems/ruby-2.6.3/bin/pod:23:in `<main>'
/Users/cloud/.rvm/gems/ruby-2.6.3/bin/ruby_executable_hooks:24:in `eval'
/Users/cloud/.rvm/gems/ruby-2.6.3/bin/ruby_executable_hooks:24:in `<main>'
```

**Comment by dnkoutso** [comment]
hmm interesting, I will try with the link you provided.

**Comment by dnkoutso** [comment]
do you happen to have an open source pod or something that i can download? The one you specified I do not have access so I cant verify.

**Comment by dnkoutso** [comment]
it seems we are catching an exception and re-throwing but we are missing the underlying message:

```
          begin
            download_result = Downloader.download(download_request, target, :can_cache => can_cache)
          rescue Pod::DSLError => e
            raise Informative, "Failed to load '#{name}' podspec: #{e.message}"
          rescue => e
            raise Informative, "Failed to download '#{name}': #{e.message}"
          end
```

Can you also run with `pod install --verbose` and see if there is a separate stacktrace printed?

**Comment by dnkoutso** [comment]
Alternatively please modify your `/Users/cloud/.rvm/gems/ruby-2.6.3/gems/cocoapods-1.10.2/lib/cocoapods/external_sources/abstract_external_source.rb:120:in `rescue in block in pre_download` file and add a `puts e` right below the `rescue => e` line on 119 and before the `raise` line. This might give us more insight about what the underlying error is.

**Comment by Buju77** [comment]
> do you happen to have an open source pod or something that i can download? The one you specified I do not have access so I cant verify.

sry, I can't provide that. this is a private repo from the company. the blub.com is just a redacted url 😅 

> Alternatively please modify your `/Users/cloud/.rvm/gems/ruby-2.6.3/gems/cocoapods-1.10.2/lib/cocoapods/external_sources/abstract_external_source.rb` file inside the `pre_download` method and add a `puts e` right below the `rescue => e` line on 119 and before the `raise` line on 120. This might give us more insight about what the underlying error is.

ok cool thx. I'll try that in the morning when I'm in the office again. The weird thing is, the error is only happening in this one repo. If we do the same thing in other repo, it works fine. So definitely interesting to see what the underlying error is.

**Comment by Buju77** [comment]
ok so `pod install --verbose` also gives the same stacktrace:

```
/Users/cloud/.rvm/gems/ruby-2.6.3/gems/cocoapods-1.10.2/lib/cocoapods/external_sources/abstract_external_source.rb:120:in `rescue in block in pre_download'
/Users/cloud/.rvm/gems/ruby-2.6.3/gems/cocoapods-1.10.2/lib/cocoapods/external_sources/abstract_external_source.rb:115:in `block in pre_download'
/Users/cloud/.rvm/gems/ruby-2.6.3/gems/cocoapods-1.10.2/lib/cocoapods/user_interface.rb:86:in `titled_section'
/Users/cloud/.rvm/gems/ruby-2.6.3/gems/cocoapods-1.10.2/lib/cocoapods/external_sources/abstract_external_source.rb:113:in `pre_download'
/Users/cloud/.rvm/gems/ruby-2.6.3/gems/cocoapods-1.10.2/lib/cocoapods/external_sources/downloader_source.rb:13:in `fetch'
/Users/cloud/.rvm/gems/ruby-2.6.3/gems/cocoapods-1.10.2/lib/cocoapods/installer/analyzer.rb:989:in `fetch_external_source'
/Users/cloud/.rvm/gems/ruby-2.6.3/gems/cocoapods-1.10.2/lib/cocoapods/installer/analyzer.rb:968:in `block (2 levels) in fetch_external_sources'
/Users/cloud/.rvm/gems/ruby-2.6.3/gems/cocoapods-1.10.2/lib/cocoapods/installer/analyzer.rb:967:in `each'
/Users/cloud/.rvm/gems/ruby-2.6.3/gems/cocoapods-1.10.2/lib/cocoapods/installer/analyzer.rb:967:in `block in fetch_external_sources'
/Users/cloud/.rvm/gems/ruby-2.6.3/gems/cocoapods-1.10.2/lib/cocoapods/user_interface.rb:64:in `section'
/Users/cloud/.rvm/gems/ruby-2.6.3/gems/cocoapods-1.10.2/lib/cocoapods/installer/analyzer.rb:966:in `fetch_external_sources'
/Users/cloud/.rvm/gems/ruby-2.6.3/gems/cocoapods-1.10.2/lib/cocoapods/installer/analyzer.rb:117:in `analyze'
/Users/cloud/.rvm/gems/ruby-2.6.3/gems/cocoapods-1.10.2/lib/cocoapods/installer.rb:414:in `analyze'
/Users/cloud/.rvm/gems/ruby-2.6.3/gems/cocoapods-1.10.2/lib/cocoapods/installer.rb:239:in `block in resolve_dependencies'
/Users/cloud/.rvm/gems/ruby-2.6.3/gems/cocoapods-1.10.2/lib/cocoapods/user_interface.rb:64:in `section'
/Users/cloud/.rvm/gems/ruby-2.6.3/gems/cocoapods-1.10.2/lib/cocoapods/installer.rb:238:in `resolve_dependencies'
/Users/cloud/.rvm/gems/ruby-2.6.3/gems/cocoapods-1.10.2/lib/cocoapods/installer.rb:160:in `install!'
/Users/cloud/.rvm/gems/ruby-2.6.3/gems/cocoapods-1.10.2/lib/cocoapods/command/install.rb:52:in `run'
/Users/cloud/.rvm/gems/ruby-2.6.3/gems/claide-1.0.3/lib/claide/command.rb:334:in `run'
/Users/cloud/.rvm/gems/ruby-2.6.3/gems/cocoapods-1.10.2/lib/cocoapods/command.rb:52:in `run'
/Users/cloud/.rvm/gems/ruby-2.6.3/gems/cocoapods-1.10.2/bin/pod:55:in `<top (required)>'
/Users/cloud/.rvm/gems/ruby-2.6.3/bin/pod:23:in `load'
/Users/cloud/.rvm/gems/ruby-2.6.3/bin/pod:23:in `<main>'
/Users/cloud/.rvm/gems/ruby-2.6.3/bin/ruby_executable_hooks:24:in `eval'
/Users/cloud/.rvm/gems/ruby-2.6.3/bin/ruby_executable_hooks:24:in `<main>'
```

**Comment by Buju77** [comment]
> Alternatively please modify your `/Users/cloud/.rvm/gems/ruby-2.6.3/gems/cocoapods-1.10.2/lib/cocoapods/external_sources/abstract_external_source.rb` file inside the `pre_download` method and add a `puts e` right below the `rescue => e` line on 119 and before the `raise` line on 120. This might give us more insight about what the underlying error is.

just tried this, but unfortunately also just prints:

```
can't modify frozen String
```

**Comment by Buju77** [comment]
ok, so I've tracked down the line that crashes, which is the `Downloader.preprocess_options(request.params)` in downloader.rb#132

I've added bunch of puts for debugging in that method:

```
    def self.preprocess_request(request)
      puts "============= DEBUG =============== creating Request.new() now ..."
      puts "============= DEBUG =============== spec: #{request.spec}"
      puts "============= DEBUG =============== real: #{request.released_pod?}"
      puts "============= DEBUG =============== name: #{request.name}"
      puts "============= DEBUG =============== para: #{request.params}"
      # puts "============= DEBUG =============== prep: #{Downloader.preprocess_options(request.params)}" # would crashes here

      begin
        crashing_params = Downloader.preprocess_options(request.params)
      rescue => e
        puts "============= DEBUG =============== error: #{e}"
      end

      req = Request.new(
        :spec => request.spec,
        :released => request.released_pod?,
        :name => request.name,
        :params => Downloader.preprocess_options(request.params))
      puts "============= DEBUG =============== after creating"
      req
    end
```

and this is the output:

```
Fetching external sources
-> Pre-downloading: 'profile' from `ssh://git@blub.com:7999/profile.git`, branch `master`
============= DEBUG =============== preprocess_request
============= DEBUG =============== creating Request.new() now ...
============= DEBUG =============== spec:
============= DEBUG =============== real: false
============= DEBUG =============== name: profile
============= DEBUG =============== para: {:git=>"ssh://git@blub.com:7999/profile.git", :branch=>"master"}
  $ /usr/bin/git ls-remote ssh://git@stash.runtastic.com:7999/profile.git master
  862052ef4a1df40d1d801a341b280eec78c9137d	refs/heads/master
============= DEBUG =============== error: can't modify frozen String
```

Where is `Downloader.preprocess_options()` implemented? I couldn't find it in the repo to add more debug puts 🤔

**Comment by Buju77** [comment]
so, after some further digging, I finally found the line that causes this error. It is actually a new line that was recently introduced in cocoapods-downloader 1.5.0 release 7 days ago:

https://github.com/CocoaPods/cocoapods-downloader/compare/1.4.0...1.5.0#diff-10b883c026b3c31048ba1993c70a5c8975ed81384dc1f19182c224cada97648c

```ruby
encoded_branch_name = branch_name.force_encoding(Encoding::ASCII_8BIT)
```

This line causes the "can't modify frozen string" exception

**Comment by Buju77** [comment]
lol seems like issue is fixed by using `encode` method instead of `force_encoding` ✅  🎉 

```ruby
encoded_branch_name = branch_name.encode(Encoding::ASCII_8BIT)
```

**Comment by Buju77** [comment]
I've opened a PR now. Would really appreciate it if somebody could release the bugfix pretty soon :)

**Comment by dnkoutso** [comment]
awesome! I will take a look, we can ship cocoapods-downloader 1.5.1 for this. I will help you land the PR.

**Comment by dnkoutso** [comment]
This was fixed in cocoapods-downloader and shipped with version 1.5.1. Please use `gem update cocoapods-downloader` if you encounter this error.
